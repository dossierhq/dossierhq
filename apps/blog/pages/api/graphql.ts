import { ok } from '@jonasb/datadata-core';
import type { SessionGraphQLContext } from '@jonasb/datadata-graphql';
import { GraphQLSchemaGenerator } from '@jonasb/datadata-graphql';
import type { ExecutionResult, GraphQLSchema } from 'graphql';
import { graphql } from 'graphql';
import type { NextApiRequest, NextApiResponse } from 'next';
import { handleRequest, sendMethodNotAllowedError } from '../../utils/HandlerUtils';
import { getServerConnection, getSessionContextForRequest } from '../../utils/ServerUtils';

let graphQLSchema: GraphQLSchema | null = null;

export default async function graphQlHandler(
  req: NextApiRequest,
  res: NextApiResponse<ExecutionResult>
): Promise<void> {
  if (req.method !== 'POST') {
    sendMethodNotAllowedError(res, ['POST']);
    return;
  }

  await handleRequest(res, async () => {
    const { server, schema } = await getServerConnection();
    const authResult = await getSessionContextForRequest(server, req);
    if (authResult.isError()) return authResult;
    const { adminClient, publishedClient } = authResult.value;

    const context: SessionGraphQLContext = {
      adminClient: ok(adminClient),
      publishedClient: ok(publishedClient),
    };

    if (!graphQLSchema) {
      graphQLSchema = new GraphQLSchemaGenerator({
        adminSchema: schema,
        publishedSchema: schema.toPublishedSchema(),
      }).buildSchema();
    }

    const { query, variables, operationName } = req.body;
    const result = await graphql({
      schema: graphQLSchema,
      source: query,
      contextValue: context,
      variableValues: variables,
      operationName,
    });

    return ok(result);
  });
}
