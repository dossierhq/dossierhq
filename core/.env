DATABASE_URL=postgres://coretestuser:coretestpass@localhost:5432/datadata-core
