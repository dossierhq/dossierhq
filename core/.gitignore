/node_modules/
/docker-compose.dockest-generated.yml
