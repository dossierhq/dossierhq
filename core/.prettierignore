/package-lock.json
/tsconfig.json
/CHANGELOG.md
