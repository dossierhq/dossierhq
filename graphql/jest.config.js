module.exports = {
  displayName: 'graphql',
  preset: 'ts-jest',
  testEnvironment: 'node',
  globalSetup: './test/setup.ts',
  globalTeardown: './test/teardown.ts',
};
