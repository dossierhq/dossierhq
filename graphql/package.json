{
  "name": "@jonasb/datadata-graphql",
  "version": "0.1.1-7",
  "description": "",
  "main": "./lib/index.js",
  "types": "./lib/index.d.ts",
  "exports": {
    ".": "./lib/index.js"
  },
  "scripts": {
    "build": "heft build",
    "build:watch": "heft build --watch",
    "ci:check": "npm run test && npm run prettier:check",
    "clean": "heft clean",
    "db:cli": "./scripts/db-cli.sh",
    "db:migrate": "./scripts/db-migrate.sh",
    "eslint:check": "eslint .",
    "eslint:fix": "eslint --fix .",
    "prettier:check": "prettier --check .",
    "prettier:fix": "prettier --write .",
    "test": "heft test",
    "test:update-snapshot": "heft test --update-snapshots",
    "test:watch": "heft test --watch",
    "ts:check": "tsc --noEmit",
    "ts:check:watch": "tsc --noEmit --watch",
    "upgrade-dependencies": "npx npm-check-updates -u --packageFile ./package.json",
    "upgrade-dependencies:check": "npx npm-check-updates"
  },
  "author": "",
  "license": "ISC",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/jonasb/datadata.git"
  },
  "devDependencies": {
    "@jonasb/datadata-postgres": "0.1.1-7",
    "@jonasb/datadata-server": "0.1.1-7",
    "@jonasb/datadata-testing-utils": "0.1.1-7",
    "@rushstack/heft": "~0.34.8",
    "@rushstack/heft-jest-plugin": "~0.1.14",
    "@types/heft-jest": "1.0.2",
    "@types/uuid": "^8.3.0",
    "@typescript-eslint/eslint-plugin": "^4.24.0",
    "@typescript-eslint/parser": "^4.24.0",
    "eslint": "^7.26.0",
    "prettier": "2.3.0",
    "typescript": "^4.3.5",
    "uuid": "^8.3.2"
  },
  "dependencies": {
    "@jonasb/datadata-core": "0.1.1-7",
    "graphql": "~15.5.0"
  }
}
