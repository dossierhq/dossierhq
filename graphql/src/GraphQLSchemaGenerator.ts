import type { Schema } from '@datadata/core';
import {
  GraphQLEnumType,
  GraphQLID,
  GraphQLInterfaceType,
  GraphQLNonNull,
  GraphQLObjectType,
  GraphQLSchema,
  GraphQLString,
  isInterfaceType,
} from 'graphql';
import type { GraphQLEnumValueConfigMap, GraphQLNamedType, GraphQLSchemaConfig } from 'graphql';

export class GraphQLSchemaGenerator {
  #types: GraphQLNamedType[] = [];

  constructor(readonly schema: Schema) {}

  addType(type: GraphQLNamedType): void {
    if (this.#types.find((x) => x.name === type.name)) {
      throw new Error(`Type with name ${type.name} already exists`);
    }
    this.#types.push(type);
  }

  getType(name: string): GraphQLNamedType {
    const type = this.#types.find((x) => x.name === name);
    if (!type) {
      throw new Error(`Type with name ${name} doesn't exist`);
    }
    return type;
  }

  getInterface(name: string): GraphQLInterfaceType {
    const type = this.getType(name);
    if (isInterfaceType(type)) {
      return type;
    }
    throw new Error(`Type ${name} is not an interface`);
  }

  addSupportingTypes(): void {
    // Node
    this.addType(
      new GraphQLInterfaceType({
        name: 'Node',
        fields: {
          id: { type: new GraphQLNonNull(GraphQLID) },
        },
      })
    );

    // EntityType
    const entityTypeEnumValues: GraphQLEnumValueConfigMap = {};
    for (const typeName of Object.keys(this.schema.spec.entityTypes)) {
      entityTypeEnumValues[typeName] = {};
    }
    if (Object.keys(entityTypeEnumValues).length === 0) {
      // Can't create EntityType with no enum values
      return;
    }
    this.addType(
      new GraphQLEnumType({
        name: 'EntityType',
        values: entityTypeEnumValues,
      })
    );

    // Entity
    this.addType(
      new GraphQLInterfaceType({
        name: 'Entity',
        interfaces: [this.getInterface('Node')],
        fields: {
          id: { type: new GraphQLNonNull(GraphQLID) },
          _type: { type: new GraphQLNonNull(this.getType('EntityType')) },
          _name: { type: new GraphQLNonNull(GraphQLString) },
        },
      })
    );
  }

  buildSchemaConfig<TSource, TContext>(): GraphQLSchemaConfig {
    this.addSupportingTypes();
    const queryType = new GraphQLObjectType<TSource, TContext>({
      name: 'Query',
      fields: {
        hello: {
          type: GraphQLString,
          resolve() {
            return 'world';
          },
        },
      },
    });
    return { query: queryType, types: this.#types };
  }

  buildSchema<TSource, TContext>(): GraphQLSchema {
    return new GraphQLSchema(this.buildSchemaConfig<TSource, TContext>());
  }
}
