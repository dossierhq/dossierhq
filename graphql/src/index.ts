export type { SessionGraphQLContext } from './GraphQLSchemaGenerator';

export { GraphQLSchemaGenerator } from './GraphQLSchemaGenerator';
