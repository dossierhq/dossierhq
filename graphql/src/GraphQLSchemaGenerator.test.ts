import { EntityFieldType, Schema } from '@datadata/core';
import type { SchemaSpecification } from '@datadata/core';
import { printSchema } from 'graphql';
import { GraphQLSchemaGenerator } from './GraphQLSchemaGenerator';

async function describeGeneratedSchema(schemaSpec: SchemaSpecification) {
  const generator = new GraphQLSchemaGenerator(new Schema(schemaSpec));
  const graphQLSchema = generator.buildSchema();
  return printSchema(graphQLSchema);
}

describe('Empty schema spec', () => {
  const schemaSpec = { entityTypes: {} };
  test('Generated QL schema', async () => {
    const result = await describeGeneratedSchema(schemaSpec);
    expect(result).toMatchSnapshot();
  });
});

describe('One empty entity type schema spec', () => {
  const schemaSpec = { entityTypes: { Foo: { fields: [] } } };
  test('Generated QL schema', async () => {
    const result = await describeGeneratedSchema(schemaSpec);
    expect(result).toMatchSnapshot();
  });
});

describe('Two entity types schema spec', () => {
  const schemaSpec = {
    entityTypes: {
      Foo: { fields: [{ name: 'fooField', type: EntityFieldType.String }] },
      Bar: {
        fields: [
          { name: 'barField1', type: EntityFieldType.String },
          { name: 'barField2', type: EntityFieldType.String },
        ],
      },
    },
  };
  test('Generated QL schema', async () => {
    const result = await describeGeneratedSchema(schemaSpec);
    expect(result).toMatchSnapshot();
  });
});
