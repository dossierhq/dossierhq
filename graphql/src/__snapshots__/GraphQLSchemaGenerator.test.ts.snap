// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Empty schema spec Generated QL schema 1`] = `
Object {
  "data": Object {
    "__schema": Object {
      "types": Array [
        Object {
          "enumValues": null,
          "fields": Array [
            Object {
              "name": "id",
              "type": Object {
                "kind": "NON_NULL",
                "name": null,
                "ofType": Object {
                  "name": "ID",
                },
              },
            },
          ],
          "interfaces": Array [],
          "kind": "INTERFACE",
          "name": "Node",
        },
        Object {
          "enumValues": null,
          "fields": Array [
            Object {
              "name": "hello",
              "type": Object {
                "kind": "SCALAR",
                "name": "String",
                "ofType": null,
              },
            },
          ],
          "interfaces": Array [],
          "kind": "OBJECT",
          "name": "Query",
        },
      ],
    },
  },
}
`;

exports[`One empty entity type schema spec Generated QL schema 1`] = `
Object {
  "data": Object {
    "__schema": Object {
      "types": Array [
        Object {
          "enumValues": null,
          "fields": Array [
            Object {
              "name": "id",
              "type": Object {
                "kind": "NON_NULL",
                "name": null,
                "ofType": Object {
                  "name": "ID",
                },
              },
            },
          ],
          "interfaces": Array [],
          "kind": "INTERFACE",
          "name": "Node",
        },
        Object {
          "enumValues": Array [
            Object {
              "name": "Foo",
            },
          ],
          "fields": null,
          "interfaces": null,
          "kind": "ENUM",
          "name": "EntityType",
        },
        Object {
          "enumValues": null,
          "fields": Array [
            Object {
              "name": "id",
              "type": Object {
                "kind": "NON_NULL",
                "name": null,
                "ofType": Object {
                  "name": "ID",
                },
              },
            },
            Object {
              "name": "_type",
              "type": Object {
                "kind": "NON_NULL",
                "name": null,
                "ofType": Object {
                  "name": "EntityType",
                },
              },
            },
            Object {
              "name": "_name",
              "type": Object {
                "kind": "NON_NULL",
                "name": null,
                "ofType": Object {
                  "name": "String",
                },
              },
            },
          ],
          "interfaces": Array [
            Object {
              "name": "Node",
            },
          ],
          "kind": "INTERFACE",
          "name": "Entity",
        },
        Object {
          "enumValues": null,
          "fields": Array [
            Object {
              "name": "hello",
              "type": Object {
                "kind": "SCALAR",
                "name": "String",
                "ofType": null,
              },
            },
          ],
          "interfaces": Array [],
          "kind": "OBJECT",
          "name": "Query",
        },
      ],
    },
  },
}
`;
